Added some preliminary docs (#7)
Shopify/mybench: docs/detailed-design-doc.rst

the events and an overall event rate. In the figure, all three workloads are
configured to be executed with two parallel ``BenchmarkWorker``\s. The splitting
of workload into multiple parallel workers is useful for simulating scenarios
where many connections are sending queries to the database simultaneously.
``BenchmarkWorker``\s are implemented as goroutines each with its own connection
to the database. The overall event rate is equally divided between all
``BenchmarkWorker``\s. Each ``BenchmarkWorker`` embeds a ``Looper`` object,
which precisely controls the event rate (see `Rate control via
"temporal-discretization" looper`_), and an ``OnlineHistogram`` object, which
stores the throughput and latency statistics during the benchmark run. Data
held inside the ``OnlineHistogram`` objects are periodically read by the
``DataLogger`` (see `High-performance data logging via double buffering`_).
Once the data is read, the ``DataLogger`` aggregates the throughput and latency
statistics across all ``BenchmarkWorker``\s and all ``Workload``\s into
per-workload and overall throughput and latency statistics. These values are
then written into a SQLite database.

> Users developing a custom benchmark do not need to be concerned with the

Reviewer: Consider splitting this section into a usage section rather then leaving it in the "internals" section.

Author: I'm going to leave this here as I don't have an usage section.

> internal details of mybench. To define a ``Workload``, the user simply needs to

Reviewer: ```suggestion
internal details of mybench. To define a ``Workload``, the user needs to
```

define a struct implementing the ``WorkloadInterface`` interface, which
requires the definition of the ``Event`` callback, a ``Config`` method
returning the configurations of the workload (such as the event rate and number
of parallel workers), and a ``NewContextData`` method that constructs a
custom, thread-local data structure that could be useful for the workload to
store thread-local state. Each custom benchmark is tied together with a
``BenchmarkApp`` which constructs and owns the ``Benchmark`` object. Workloads
defined by the user can be added to the ``Benchmark`` through the
``BenchmarkApp`` object.

> Finally, the ``BenchmarkApp`` also handles other administrative duties common

Reviewer: ```suggestion

The ``BenchmarkApp`` also handles other administrative duties common
```

between all custom-defined user benchmarks, such as parsing the command line
flags and setting an HTTP server where live throughput and latency statistics
can be monitored.

> While mybench is primarily designed to benchmark MySQL, its architecture is

Reviewer: I suggest dropping this paragraph.

flexible enough to benchmark arbitrary database and non-database systems.
However, these are not uses cases that have been thoroughly tested for now.

Rate control via "temporal-discretization" looper
-------------------------------------------------

A common approach to discover the limits of the database is by performing a
sequence of benchmarks with increasing event rates. When the database is

> overloaded, the event rate will plateau and the latency will stagnate. To

Reviewer: ```suggestion
overloaded, the event rate will plateau or decline and the latency will increase. To
```

Once the data is collected, the per-workload and aggregate throughput and
latency data are written into a SQLite database on disk. Each benchmark run
occupies a single table in the SQLite database. Meta information about the
benchmark run, such as the benchmark start and stop time, is written to a table
named ``meta``. The choice of a SQLite database allows data from a series of
benchmark runs to be stored in a single file, which can simplify the transport,
storage, and post-processing of the data.

.. [HDRHIST01] http://hdrhistogram.org/

.. [#fsnapshot] Throughput is calculated as number of events divided by the
   elapsed time. Number of events is calculated by summing the event count in
   all ``OnlineHistogram`` instances. The elapsed time is calculated as the time
   difference between two subsequent sampling of all ``OnlineHistogram``
   instances. If the time required to read all ``OnlineHistogram`` instances is
   a large fraction of the elapsed time between two subsequent
   ``OnlineHistogram`` sampling, this can introduce significant bias in the
   calculated throughput, as both numerator and the denominator will be affected
   by the passage of time.

> Random data generation

Reviewer: ```suggestion
Data generation
```

----------------------

> Database benchmarks require the generation of random data. Similar to other

Reviewer: ```suggestion
Database benchmarks require the generation of data. Similar to other
```

   a large fraction of the elapsed time between two subsequent
   ``OnlineHistogram`` sampling, this can introduce significant bias in the
   calculated throughput, as both numerator and the denominator will be affected
   by the passage of time.

Data generation
----------------------

Database benchmarks require the generation of data. Similar to other
benchmark systems, mybench also provides a number of builtin data generators
that can generate data with different types and distributions. All data
generators implement an interface two methods: (1) one for the generation of
"new" data to be written to the database and (2) one for sampling data
that already exists on the database such that the data can be used in the
``WHERE`` clause of a ``SELECT`` statement. At this time, the data generator
implemented within mybench are relatively primitive. For example, accurately
sampling existing data can be memory and time intensive and mybench does not
implement a fully-correct version of the existing data sampling algorithm at
this time. Instead, data sampling is done in a best-effort manner to ensure
the high performance and small resource footprint of mybench is not
compromised.

> The random data generators implement in mybench uses the standard ``rand``

Reviewer 2: ```suggestion
The random data generators implemented in mybench use the standard ``rand``
```
